### ISSUE TITLE
SSL Certificat verification failure

### ISSUE THREAD

**Issue by blackcobra1973** [dnaeon/py-vpoller#113]
Hello,

Since today I receive the following in my log files:  

```bash
[2014-12-28 16:53:39,911 - INFO/VPollerWorker-2] Connecting vSphere Agent to vSphere
[2014-12-28 16:53:39,912 - INFO/VPollerWorker-2] Starting new HTTPS connection (1): vSphere
[2014-12-28 16:53:39,922 - ERROR/VPollerWorker-2] Cannot connect to vSphere: (vim.fault.HostConnectFault) {
   dynamicType = <unset>,
   dynamicProperty = (vmodl.DynamicProperty) [],
   msg = '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)',
   faultCause = <unset>,
   faultMessage = (vmodl.LocalizableMessage) []
}
```
This happened after a reboot of my zabbix host.  

This did not happen before, could you check ? 

Thanks in advance

Kurt

**Comment by willfroning** [comment]
Hello Marin & Kurt,

This is caused by upgrading python to 2.7.9. I haven't found a solution yet, but reverting to 2.7.8 fixes this. Release notes from [python 2.7.9](https://www.python.org/downloads/release/python-279/):

    * HTTPS certificate validation using the system's certificate store is now enabled by default. See PEP 476 for details.

Thanks,
Will

**Comment by dnaeon** [comment]
Hey Will,

Thanks for mentioning this!

I can provide a patch with a fix soon.

Thanks again,
Marin

**Comment by Robbinton** [comment]
Hello Marin & Kurt,
I faced with the same problem.
Gentoo linux, Python 2.7.9
Now ESXi host certificates checked against openssl on local PC.
I have added ca signed certificate to one of my esx host and add CA certificates to openssl. This ESX host is connected OK.
But most of my ESX host has selfSigned certificates. Is there a way to tell pyvmomi do not check these certificates?

**Comment by hartsock** [comment]
Hi folks, I just read [PEP 0466](https://www.python.org/dev/peps/pep-0466/) and I wonder if it's not pertinent.

**Comment by bicofino** [comment]
I got the same error.

/usr/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:769: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
[2015-03-19 18:12:35,942 - ERROR/VPollerWorker-4] Cannot connect to 172.16.20.20: (vim.fault.HostConnectFault) {
   dynamicType = <unset>,
   dynamicProperty = (vmodl.DynamicProperty) [],
   msg = '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)',
   faultCause = <unset>,
   faultMessage = (vmodl.LocalizableMessage) []

**Comment by bicofino** [comment]
Looks like it's because of Python 2.7.9
http://www.afpy.org/doc/python/2.7/whatsnew/2.7.html

**Comment by bicofino** [comment]
No issue with Python 2.7.8

**Comment by paul999** [comment]
Does anyone have a fix for this? I am having the same problem with the current version of vpoller/pyvmomi and the debian jessie version of python.

**Comment by dnaeon** [comment]
@paul999 I've just pushed a fixed for to the latest vPoller release. Please upgrade to vPoller 0.5.5, which comes with a fix for this issue.

**Comment by paul999** [comment]
Yep, indeed seems to work. Thanks a lot!

**Comment by tomoshokas** [comment]
Sorry I a newbie, tried to resolve by myself, but could not ;(;( 

I saw that issue with verification of ssl certificate should be solved in vpoller version 5.5 and I should not get it on version 5.9, but it is not a case. As well it is strange that verification is by default enabled only from python version 2.7.9 but on centos 7.1 it is still version 2.7.5.

I do have an issue with vpoller  to connect to vcenter  6.0.1 U1 vcsa. Error output is at the end from vpoller.

Configuration is made according you instructions on centos 7.1 64 bits python is default to centos 2.7.5

# pip freeze | egrep 'pyvmomi|vpoller|vconnector|requests'
pyvmomi==6.0.0
requests==2.8.1
vconnector==0.4.1
vpoller==0.5.9

[root@s-eci-monitor opt]# vpoller-proxy start &
[1] 11550
[root@s-eci-monitor opt]# [2015-12-04 10:35:11,466 - INFO/MainProcess] Starting Proxy Manager [0.5.9 release]
[2015-12-04 10:35:11,467 - INFO/MainProcess] Creating Proxy Manager sockets
[2015-12-04 10:35:11,468 - INFO/MainProcess] Starting Proxy process
[2015-12-04 10:35:11,469 - INFO/MainProcess] Proxy Manager is ready and running
[2015-12-04 10:35:11,469 - INFO/VPollerProxy-1] Proxy process is starting
[2015-12-04 10:35:11,469 - INFO/VPollerProxy-1] Creating Proxy process sockets
[2015-12-04 10:35:11,471 - INFO/VPollerProxy-1] Proxy process is ready and running


[root@s-eci-monitor opt]# vpoller-worker start &
[3] 11612
[2]   Done                    vpoller-worker start
[root@s-eci-monitor opt]# [2015-12-04 10:36:28,674 - INFO/MainProcess] Starting Worker Manager [0.5.9 release]
[2015-12-04 10:36:28,675 - INFO/MainProcess] Starting Worker processes
[2015-12-04 10:36:28,676 - INFO/MainProcess] Concurrency: 2 (processes)
[2015-12-04 10:36:28,677 - INFO/VPollerWorker-1] Worker process is starting
[2015-12-04 10:36:28,678 - INFO/VPollerWorker-1] Loading task module vpoller.vsphere.tasks
[2015-12-04 10:36:28,678 - INFO/MainProcess] Worker Manager is ready and running
[2015-12-04 10:36:28,680 - INFO/VPollerWorker-2] Worker process is starting
[2015-12-04 10:36:28,680 - INFO/VPollerWorker-2] Loading task module vpoller.vsphere.tasks
[2015-12-04 10:36:28,680 - INFO/VPollerWorker-1] Loading helper module vpoller.helpers.zabbix
[2015-12-04 10:36:28,681 - INFO/VPollerWorker-1] Loading helper module vpoller.helpers.czabbix
[2015-12-04 10:36:28,681 - INFO/VPollerWorker-1] Creating Worker sockets
[2015-12-04 10:36:28,682 - INFO/VPollerWorker-2] Loading helper module vpoller.helpers.zabbix
[2015-12-04 10:36:28,682 - INFO/VPollerWorker-2] Loading helper module vpoller.helpers.czabbix
[2015-12-04 10:36:28,683 - INFO/VPollerWorker-2] Creating Worker sockets
[2015-12-04 10:36:28,683 - INFO/VPollerWorker-1] Created vSphere Agent for 158.166.219.34
[2015-12-04 10:36:28,683 - INFO/VPollerWorker-1] Worker process is ready and running
[2015-12-04 10:36:28,684 - INFO/VPollerWorker-2] Created vSphere Agent for 158.166.219.34
[2015-12-04 10:36:28,685 - INFO/VPollerWorker-2] Worker process is ready and running



I tried to use the script which you suggested to protofab in vpoller-client unknown or missing agent name #142 :

from vconnector.core import VConnector                                                                                                                                                                        
agent = VConnector(user='*******', pwd='********', host='158.166.2119.34')
hosts = agent.get_host_view().view
print hosts

I get following error:

 python vcenter-test.py
ERROR:root:Cannot connect to 158.166.219.34: bad handshake: Error([('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)
Traceback (most recent call last):
  File "vcenter-test.py", line 3, in <module>
    hosts = agent.get_host_view().view
  File "/usr/lib/python2.7/site-packages/vconnector/core.py", line 192, in get_host_view
    obj_type=[pyVmomi.vim.HostSystem]
  File "/usr/lib/python2.7/site-packages/vconnector/core.py", line 320, in get_container_view
    container = self.si.content.rootFolder
  File "/usr/lib/python2.7/site-packages/vconnector/core.py", line 106, in si
    self.connect()
  File "/usr/lib/python2.7/site-packages/vconnector/core.py", line 141, in connect
    pwd=self.pwd
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 663, in SmartConnect
    sslContext)
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 552, in __FindSupportedVersion
    sslContext)
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 472, in __GetServiceVersionDescription
    tree = __GetElementTreeFromUrl(url, sslContext)
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 440, in __GetElementTreeFromUrl
    sock = requests.get(url)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 69, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 433, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: bad handshake: Error([('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)

[root@s-eci-monitor opt]# vpoller-client --method event.latest --vsphere-host 158.166.219.34
[2015-12-04 10:38:31,246 - INFO/VPollerWorker-1] [158.166.219.34] Retrieving latest registered event
[2015-12-04 10:38:31,246 - INFO/VPollerWorker-1] Connecting vSphere Agent to 158.166.219.34
[2015-12-04 10:38:31,257 - INFO/VPollerWorker-1] Starting new HTTPS connection (1): 158.166.219.34
[2015-12-04 10:38:31,275 - ERROR/VPollerWorker-1] Cannot connect to 158.166.219.34: bad handshake: Error([('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)
[2015-12-04 10:38:31,277 - WARNING/VPollerWorker-1] Task event.latest failed: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/vpoller/task/decorators.py", line 55, in wrapper
    result = fn(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/vpoller/vsphere/tasks.py", line 592, in event_latest
    e = agent.si.content.eventManager.latestEvent.fullFormattedMessage
  File "/usr/lib/python2.7/site-packages/vconnector/core.py", line 106, in si
    self.connect()
  File "/usr/lib/python2.7/site-packages/vconnector/core.py", line 141, in connect
    pwd=self.pwd
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 663, in SmartConnect
    sslContext)
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 552, in __FindSupportedVersion
    sslContext)
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 472, in __GetServiceVersionDescription
    tree = __GetElementTreeFromUrl(url, sslContext)
  File "/usr/lib/python2.7/site-packages/pyVim/connect.py", line 440, in __GetElementTreeFromUrl
    sock = requests.get(url)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 69, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 433, in send
    raise SSLError(e, request=request)
SSLError: bad handshake: Error([('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)

{"msg": "Task event.latest failed", "traceback": "Traceback (most recent call last):\n  File \"/usr/lib/python2.7/site-packages/vpoller/task/decorators.py\", line 55, in wrapper\n    result = fn(*args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/vpoller/vsphere/tasks.py\", line 592, in event_latest\n    e = agent.si.content.eventManager.latestEvent.fullFormattedMessage\n  File \"/usr/lib/python2.7/site-packages/vconnector/core.py\", line 106, in si\n    self.connect()\n  File \"/usr/lib/python2.7/site-packages/vconnector/core.py\", line 141, in connect\n    pwd=self.pwd\n  File \"/usr/lib/python2.7/site-packages/pyVim/connect.py\", line 663, in SmartConnect\n    sslContext)\n  File \"/usr/lib/python2.7/site-packages/pyVim/connect.py\", line 552, in __FindSupportedVersion\n    sslContext)\n  File \"/usr/lib/python2.7/site-packages/pyVim/connect.py\", line 472, in __GetServiceVersionDescription\n    tree = __GetElementTreeFromUrl(url, sslContext)\n  File \"/usr/lib/python2.7/site-packages/pyVim/connect.py\", line 440, in __GetElementTreeFromUrl\n    sock = requests.get(url)\n  File \"/usr/lib/python2.7/site-packages/requests/api.py\", line 69, in get\n    return request('get', url, params=params, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/requests/api.py\", line 50, in request\n    response = session.request(method=method, url=url, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/requests/sessions.py\", line 468, in request\n    resp = self.send(prep, **send_kwargs)\n  File \"/usr/lib/python2.7/site-packages/requests/sessions.py\", line 576, in send\n    r = adapter.send(request, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/requests/adapters.py\", line 433, in send\n    raise SSLError(e, request=request)\nSSLError: bad handshake: Error([('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)\n", "success": 1}

[

**Comment by dnaeon** [comment]
@tomoshokas Please update your vPoller and vConnector to the latest ones from PyPI. These releases make use of the recent pyVmomi 6.0.0 as well as using the new sslContext flag as provided by this release of pyVmomi.

* https://github.com/vmware/pyvmomi/issues/212
